src/core/thread/network_diagnostic_tlvs.hpp

```diff
@@ -0,0 +1,1307 @@
+        kMode                = 2,    ///< Mode TLV
+        kTimeout             = 3,    ///< Timeout TLV
+        kConnectivity        = 4,    ///< Connectivity TLV
+        kRoute               = 5,    ///< Routing-Table TLV
+        kLeaderData          = 6,    ///< Leader Data TLV
+        kNetworkData         = 7,    ///< Network Data TLV
+        kIPv6AddressList     = 8,    ///< IPv6 Address List TLV
+        kMacCounters         = 9,    ///< Mac Counters TLV
+        kBatteryLevel        = 14,   ///< Battery Level TLV
+        kSupplyVoltage       = 15,   ///< Supply Voltage TLV
+        kChildTable          = 16,   ///< Child Table TLV
+        kChannelPages        = 17,   ///< Channel Pages TLV
+        kInvalid             = 255,
+    };
+
+    /**
+     * This method returns the Type value.
+     *
+     * @returns The Type value.
+     *
+     */
+    Type GetType(void) const { return static_cast<Type>(mType); }
+
+    /**
+     * This method sets the Type value.
+     *
+     * @param[in]  aType  The Type value.
+     *
+     */
+    void SetType(Type aType) { mType = static_cast<uint8_t>(aType); }
+
+    /**
+     * This method returns the Length value.
+     *
+     */
+    uint8_t GetLength(void) const { return mLength; }
+
+    /**
+     * This method returns the length to be sent
+     */
+    uint8_t GetSize(void) const { return mLength + 2; }
+
+    /**
+     * This method sets the Length value.
+     *
+     * @param[in]  aLength  The Length value.
+     *
+     */
+    void SetLength(uint8_t aLength) { mLength = aLength; }
+
+    /**
+     * This static method reads the requested TLV out of @p aMessage.
+     *
+     * @param[in]   aMessage    A reference to the message.
+     * @param[in]   aType       The Type value to search for.
+     * @param[in]   aMaxLength  Maximum number of bytes to read.
+     * @param[out]  aTlv        A reference to the TLV that will be copied to.
+     *
+     * @retval kThreadError_None      Successfully copied the TLV.
+     * @retval kThreadError_NotFound  Could not find the TLV with Type @p aType.
+     *
+     */
+    static ThreadError GetTlv(const Message &aMessage, Type aType, uint16_t aMaxLength, NetworkDiagnosticTlv &aTlv);
+
+    /**
+     * This static method obtains the offset of a TLV within @p aMessage.
+     *
+     * @param[in]   aMessage    A reference to the message.
+     * @param[in]   aType       The Type value to search for.
+     * @param[out]  aOffset     A reference to the offset of the TLV.
+     *
+     * @retval kThreadError_None      Successfully copied the TLV.
+     * @retval kThreadError_NotFound  Could not find the TLV with Type @p aType.
+     *
+     */
+    static ThreadError GetOffset(const Message &aMessage, Type aType, uint16_t &aOffset);
+
+private:
+    uint8_t mType;
+    uint8_t mLength;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Extended Address TLV generation and parsing.
+ *
+ */
+class ExtMacAddressTlv: public NetworkDiagnosticTlv
```

Comment:
I'm happy to inherit and tried to do that, but gave up for it looks weird to inherit from another same level component. Is it better to create a TLV base class, so that a lot of code can be shared.